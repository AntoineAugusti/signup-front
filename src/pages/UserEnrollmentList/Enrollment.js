import React, { useCallback, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import './Enrollment.css';

import { TARGET_API_LABELS, API_ICONS } from '../../lib/api';
import { USER_STATUS_LABELS } from '../../lib/enrollment';

import ActivityFeedWrapper from './ActivityFeedWrapper';
import { getAPIStats } from '../../lib/services';

const Enrollment = ({
  id,
  user,
  events,
  target_api,
  intitule,
  description,
  status,
  updated_at,
  onSelect,
}) => {
  const [stats, setStats] = useState(null);

  useEffect(() => {
    async function fetchStats() {
      const result = await getAPIStats(target_api);
      setStats(result.data);
    }

    if (status === 'sent') {
      fetchStats();
    }
  }, [status, target_api]);

  const handleClick = useCallback(
    e => {
      onSelect(e, id, target_api);
    },
    [id, target_api, onSelect]
  );

  return (
    <div className="enrollment">
      <div className="enrollment-header">
        <div className="fs">
          {TARGET_API_LABELS[target_api]}{' '}
          {API_ICONS[target_api] && (
            <div>
              <img
                src={`/images/${API_ICONS[target_api]}`}
                alt={`logo ${TARGET_API_LABELS[target_api]}`}
              />
            </div>
          )}
        </div>
        <div className={`status ${status}`}>{USER_STATUS_LABELS[status]}</div>
      </div>

      <div className="enrollment-body">
        <div className="title">
          <div className="intitule">{intitule || 'Aucun intitulé'}</div>
          <div>
            Demande effectuée le <b>{moment(updated_at).format('L')}</b> par{' '}
            <b>{user.email}</b>
          </div>
        </div>
        {<p>{description || 'Aucune description'}</p>}

        <ActivityFeedWrapper
          averageProcessingTimeInDays={
            stats
              ? Math.round(stats.average_processing_time_in_days * 100) / 100
              : 0
          }
          status={status}
          events={events}
          updated_at={updated_at}
        />

        <div className="enrollment-footer">
          <div>
            <b>N° {id}</b>
          </div>
          <button className="button large" onClick={handleClick}>
            Consulter
          </button>
        </div>
      </div>
    </div>
  );
};

Enrollment.propTypes = {
  id: PropTypes.number.isRequired,
  user: PropTypes.shape({
    email: PropTypes.string.isRequired,
  }).isRequired,
  target_api: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  updated_at: PropTypes.string.isRequired,
  description: PropTypes.string,
  intitule: PropTypes.string,
  onSelect: PropTypes.func.isRequired,
};

export default Enrollment;
