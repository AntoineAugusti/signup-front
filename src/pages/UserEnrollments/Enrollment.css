.enrollment {
  width: 100%;
  box-shadow: 0 1px 4px 0 #C9D3DF;
  background-color: var(--white);
}

.enrollment-header {
  display: flex;
  justify-content: space-between;
  flex-flow: wrap;
  align-items: center;
  padding: 1em;
  border-bottom: 1px solid var(--light-grey);
}

.enrollment-header .fs {
  font-size: 1.25em;
  font-weight: 600;
  width: 80%;
}

.title {
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-flow: wrap;
}

.intitule {
  font-weight: 600;
  font-size: large;
}

.enrollment-body {
  padding: 1em;
}

.enrollment-footer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 2em;
}

.tag {
  padding: 0.5em 1em;
  text-align: center;
  border-radius: 8px;
  color: var(--white);
}

.tag.pending-tag {
  background-color: var(--light-grey);
  color: var(--black);
}

.tag.sent-tag {
  background-color: var(--light-blue);
}

.tag.validated-tag {
  background-color: var(--green);
}

.tag.modification_pending-tag {
  background-color: var(--orange);
}

.tag.refused-tag {
  background-color: var(--red);
}
