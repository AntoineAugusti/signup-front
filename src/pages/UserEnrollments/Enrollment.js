import React, { useCallback, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import moment from 'moment';

import './Enrollment.css';

import { hashToQueryParams } from '../../lib/utils';

import { TARGET_API_LABELS } from '../EnrollmentList';

import History from './History';

const { REACT_APP_BACK_HOST: BACK_HOST } = process.env;

export const STATUS_LABELS = {
  pending: 'Brouillon',
  modification_pending: 'Modifications demandés',
  sent: 'En cours de validation',
  validated: 'Validée',
  refused: 'Refusée',
};

export const API_ICONS = {
  franceconnect: 'logo-fc.png',
  api_droits_cnam: 'logo-cnam.jpg',
  api_impot_particulier: 'logo-dgfip.png',
  api_impot_particulier_step2: 'logo-dgfip.png',
};

const Enrollment = ({
  id,
  user,
  events,
  target_api,
  intitule,
  description,
  status,
  updated_at,
  onSelect,
}) => {
  const [data, setData] = useState(null);

  useEffect(() => {
    async function fetchData() {
      const result = await axios(
        `${BACK_HOST}/api/stats${hashToQueryParams({ target_api })}`
      );
      setData(result.data);
    }

    fetchData();
  }, [target_api]);

  const handleClick = useCallback(
    e => {
      onSelect(e, id, target_api);
    },
    [id, target_api, onSelect]
  );

  return (
    <div className="enrollment">
      <div className="enrollment-header">
        <div className="fs">
          {TARGET_API_LABELS[target_api]}{' '}
          {API_ICONS[target_api] && (
            <div>
              <img
                src={`/images/${API_ICONS[target_api]}`}
                alt={`logo ${TARGET_API_LABELS[target_api]}`}
              />
            </div>
          )}
        </div>
        <div className={`tag ${status}-tag`}>{STATUS_LABELS[status]}</div>
      </div>

      <div className="enrollment-body">
        <div className="title">
          <div className="intitule">{intitule || 'Aucun intitulé'}</div>
          <div>
            Demande effectuée le <b>{moment(updated_at).format('L')}</b> par{' '}
            <b>{user.email}</b>
          </div>
        </div>
        {<p>{description || 'Aucune description'}</p>}

        {data && (
          <History
            averageProcessingTimeInDays={
              Math.round(data.average_processing_time_in_days * 100) / 100
            }
            status={status}
            events={events}
            updated_at={updated_at}
          />
        )}

        <div className="enrollment-footer">
          <div>
            <b>N° {id}</b>
          </div>
          {status !== 'refused' && (
            <button className="button large" onClick={handleClick}>
              Consulter
            </button>
          )}
        </div>
      </div>
    </div>
  );
};

Enrollment.propTypes = {
  id: PropTypes.number.isRequired,
  user: PropTypes.shape({
    email: PropTypes.string.isRequired,
  }).isRequired,
  target_api: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  updated_at: PropTypes.string.isRequired,
  description: PropTypes.string,
  intitule: PropTypes.string,
  onSelect: PropTypes.func.isRequired,
};

export default Enrollment;
