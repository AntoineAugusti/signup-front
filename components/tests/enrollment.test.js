import ReactTestRenderer from 'react-test-renderer'
import Enrollment from '../enrollment'

describe('components | Enrollment', () => {
  describe('with enrollment that can send_application at pending state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        applicant: {
          email: 'test@test.test'
        },
        messages: [],
        state: 'pending',
        acl: {
          send_application: true // eslint-disable-line camelcase
        }
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })

  describe('with enrollment that can validate_application, refuse_application and review_application at sent state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        applicant: {
          email: 'test@test.test'
        },
        messages: [],
        state: 'sent',
        acl: {
          validate_application: true, // eslint-disable-line camelcase
          refuse_application: true, // eslint-disable-line camelcase
          review_application: true // eslint-disable-line camelcase
        }
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })

  describe('with enrollment that cannot validate_application at sent state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        messages: [],
        applicant: {
          email: 'test@test.test'
        },
        state: 'sent',
        acl: {
          validate_application: false // eslint-disable-line camelcase
        }
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })

  describe('with enrollment that cannot send_technical_inputs at validated state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        applicant: {
          email: 'test@test.test'
        },
        messages: [],
        state: 'validated',
        acl: {
          send_technical_inputs: false // eslint-disable-line camelcase
        }
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })

  describe('with enrollment that can send_technical_inputs at validated state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        applicant: {
          email: 'test@test.test'
        },
        messages: [],
        state: 'validated',
        acl: {
          send_technical_inputs: true // eslint-disable-line camelcase
        }
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })

  describe('with enrollment that can deploy_application at technical_inputs state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        applicant: {
          email: 'test@test.test'
        },
        messages: [],
        state: 'technical_inputs',
        acl: {
          deploy_application: true // eslint-disable-line camelcase
        }
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })

  describe('with enrollment at deployed state', () => {
    const props = {
      enrollment: {
        demarche: {
          intitule: 'test'
        },
        applicant: {
          email: 'test@test.test'
        },
        messages: [],
        state: 'deployed',
        acl: {}
      }
    }
    const renderer = ReactTestRenderer.create(<Enrollment {...props} />)
    it('should be defined', () => {
      expect(renderer).toBeDefined()
    })
    it('should match snapshot', () => {
      expect(renderer).toMatchSnapshot()
    })
  })
})
