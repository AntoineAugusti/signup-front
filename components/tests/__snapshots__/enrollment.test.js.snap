// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | Enrollment with enrollment at deployed state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    Déployé
  </p>
  <p />
  <div />
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;

exports[`components | Enrollment with enrollment that can deploy_application at technical_inputs state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    En attente de déploiement
  </p>
  <p />
  <div>
    <button
      className="button"
      id="submit"
      name="deploy_application"
      onClick={[Function]}
      type="submit"
    >
      Déployer l'application
    </button>
  </div>
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;

exports[`components | Enrollment with enrollment that can send_application at pending state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    Demande en attente
  </p>
  <p />
  <div>
    <button
      className="button"
      id="submit"
      name="send_application"
      onClick={[Function]}
      type="submit"
    >
      Envoyer la demande
    </button>
  </div>
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;

exports[`components | Enrollment with enrollment that can send_technical_inputs at validated state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    Demande validée
  </p>
  <p />
  <div>
    <button
      className="button"
      id="submit"
      name="send_technical_inputs"
      onClick={[Function]}
      type="submit"
    >
      Demander à entrer en production
    </button>
  </div>
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;

exports[`components | Enrollment with enrollment that can validate_application, refuse_application and review_application at sent state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    Demande envoyée
  </p>
  <p>
    <input
      name="content"
      onChange={[Function]}
      type="text"
      value=""
    />
  </p>
  <div>
    <button
      className="button"
      id="submit"
      name="refuse_application"
      onClick={[Function]}
      type="submit"
    >
      Refuser
    </button>
    <button
      className="button"
      id="submit"
      name="validate_application"
      onClick={[Function]}
      type="submit"
    >
      Valider
    </button>
    <button
      className="button"
      id="submit"
      name="review_application"
      onClick={[Function]}
      type="submit"
    >
      Demande de modifications
    </button>
  </div>
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;

exports[`components | Enrollment with enrollment that cannot send_technical_inputs at validated state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    Demande validée
  </p>
  <p />
  <div />
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;

exports[`components | Enrollment with enrollment that cannot validate_application at sent state should match snapshot 1`] = `
<li
  className="panel"
>
  <h2>
    test
  </h2>
  <em>
    test@test.test
  </em>
  <p />
  <p>
    État de la demande :  
    Demande envoyée
  </p>
  <p />
  <div />
  <div
    className="button-list"
  >
    <a
      className="button"
      href="/undefined.html?id="
      name="subscribe"
      onClick={[Function]}
    >
      Voir
    </a>
  </div>
</li>
`;
